src/subscribers/serial_subscriber.jl

```diff
@@ -0,0 +1,182 @@
+msg_block_size = 256
+serial_port_buffer_size = 1024
+
+mutable struct SerialSubscriber <: Subscriber
+    serial_port::LibSerialPort.SerialPort
+
+    name::String
+
+    read_buffer::StaticArrays.MVector{serial_port_buffer_size, UInt8}
+
+    msg_in_buffer::StaticArrays.MVector{msg_block_size, UInt8}
+    msg_in_length::Int64
+
+    flags::SubscriberFlags
+
+    function SerialSubscriber(serial_port::LibSerialPort.SerialPort;
+                              name = gensubscribername(),
+                              )
+        @catchserial(LibSerialPort.open(serial_port),
+                    "Failed to connect to serial port: `$serial_port`"
+                    )
+        # close(serial_port)
+
+        # Buffer for dumping read bytes into
+        read_buffer = StaticArrays.@MVector zeros(UInt8, serial_port_buffer_size)
+
+        # Vector written to when encoding Protobuf using COBS protocol
+        msg_in_buffer = StaticArrays.@MVector zeros(UInt8, msg_block_size)
+        msg_in_length = 0
+
+        @info "Subscribing $name on serial port"
+        new(
+            serial_port,
+            name,
+            read_buffer,
+            msg_in_buffer,
+            msg_in_length,
+            SubscriberFlags(),
+            )
+    end
+end
+
+function SerialSubscriber(port_name::String,
+                          baudrate::Int64;
+                          name = gensubscribername(),
+                          )
+    local sp
+    @catchserial(
+        begin
+            sp = LibSerialPort.open(port_name, baudrate)
+            LibSerialPort.close(sp)
+        end,
+        "Failed to connect to serial port at: `$port_name`"
+        )
+
+    return SerialSubscriber(sp; name = name)
+end
+
+
+Base.isopen(sub::SerialSubscriber) = LibSerialPort.isopen(sub.serial_port)
+Base.close(sub::SerialSubscriber) = LibSerialPort.close(sub.serial_port)
+
+
+"""
+    Base.readuntil(ard::Arduino, delim::UInt8)
+Reads byte by byte from arduinos serial port stream and copies into
+read buffer until 0x00 flag bit is encountered. Returns view into
+read buffer if found complete message and nothing otherwise.
+"""
+function Base.readuntil(sub::SerialSubscriber, delim::UInt8)
```

Comment:
Instead of returning `nothing` if it doesn't work, consider returning an empty view, that way the output type is consistent.